Add database

Apfe1Saft opened the pull request:
Sprint 10-11

Apfe1Saft committed:
update FilmController,UserController,Film,User
add data.sql,FilmDBStorage,FilmDBStorageImpl,UserDBStorage,UserDBStorageImpl,MPADBStorage,MPADBStorageImpl,GenreDBStorage,GenreDBStorageImpl,schema.sql

Apfe1Saft committed:
update FilmController,UserController,Film,User
add data.sql,FilmDBStorage,FilmDBStorageImpl,UserDBStorage,UserDBStorageImpl,MPADBStorage,MPADBStorageImpl,GenreDBStorage,GenreDBStorageImpl,schema.sql

Apfe1Saft committed:
Update README.md

Apfe1Saft committed:
Update README.md

Apfe1Saft committed:
Merge branch 'main' into add-database

Troitsa commented on pom.xml:
```diff
@@ -23,6 +23,18 @@
             <artifactId>spring-boot-starter-web</artifactId>
         </dependency>
 
+        <dependency>
```
16 строка у Вас была добавлена ранее, поэтому нет возможности ее комментировать.

В проекте используется javax.xml.bind, этого пакета нет в 11 java. Поэтому надо заменить здесь версию:
 <properties>
        <java.version>15</java.version>
    </properties>

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/dao/GenreDBStorage.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.Set;
+
+public interface GenreDBStorage {
+    public Film.Genre getGenreById(int id);
```
модификаторы в интерфейсе?)

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/dao/MPADBStorage.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.List;
+
+public interface MPADBStorage {
+    public MPA getMPAById(int id);
```
и здесь

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/model/MPA.java:
```diff
@@ -0,0 +1,50 @@
+        this.name = getMPANameById(id);
+    }
+
+    public String getMPANameById(int id) {
```
мы MPA и жанры храним в базе данных. Что если в бд сменяться id и указанные здесь id будут соответствовать другим значениям?
Верное решение - за наименованием ходить в бд

Apfe1Saft committed:
update FilmController,UserController,Film,User
add data.sql,FilmDBStorage,FilmDBStorageImpl,UserDBStorage,UserDBStorageImpl,MPADBStorage,MPADBStorageImpl,GenreDBStorage,GenreDBStorageImpl,schema.sql

Apfe1Saft committed:
update schema.sql

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check connection error

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/model/Genre.java:
```diff
@@ -0,0 +1,27 @@
+
+    public Genre(int id) {
+        this.id = id;
+        this.name = GenreController.getGenreByID(id).getName();
```
как в обычный POJO-объект мог попасть контроллер? Все "добычи" имен из базы данных - дело сервиса
достаточно полей и сеттеров/геттеров

@Getter
@Setter
@ToString
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Genre {
    @EqualsAndHashCode.Include
    private int id;
    private String name;

    public Genre(int id) {
        this.id = id;
    }

    public Genre(int id, String name) {
        this.id = id;
        this.name = name;
    }
}

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/model/MPA.java:
```diff
@@ -0,0 +1,45 @@
+@Slf4j
+@Data
+public class MPA {
+    private MPAService service;
```
здесь еще и сервис))


@Data
@Validated
public class MPA {
    @Min(1)
    @Max(7)
    int id;
    String name;

    public MPA() {
    }

    public MPA(int id) {
        this.id = id;
    }

    public MPA(int id, String name) {
        this.id = id;
        this.name = name;
    }

    public void setId(int id) {
        this.id = id;
    }
}

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/model/MPA.java:
```diff
@@ -0,0 +1,45 @@
+
+
+@Validated
+@RestController
```
entity-объект точно не должен быть рест-контроллером

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/model/MPA.java:
```diff
@@ -0,0 +1,45 @@
+@Validated
+@RestController
+@RequestMapping("/mpa")
+@Slf4j
```
логгирование такому классу тоже не нужно

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/MPADBStorageImpl.java:
```diff
@@ -0,0 +1,42 @@
+@Component
+public class MPADBStorageImpl implements MPADBStorage {
+
+    private final Logger log = LoggerFactory.getLogger(FilmDBStorageImpl.class);
```
@Override
    public MPA getMPAById(int id) {
        final String sqlQuery = "select * from MPA where MPA_ID = ?";
        final List<MPA> films = jdbcTemplate.query(sqlQuery, MPADBStorageImpl::makeMpa, id);
        if (films.size() != 1) {
            throw new ObjectNotFoundException("map id=" + id);
        }
        return films.get(0);
    }

    @Override
    public List<MPA> getAllMPA() {
        return jdbcTemplate.query("select * from MPA", MPADBStorageImpl::makeMpa);
    }

    static MPA makeMpa(ResultSet rs, int rowNum) throws SQLException {
        return new MPA(
                rs.getInt("mpa_id"),
                rs.getString("name"));
    }

Apfe1Saft committed:
check connection error

Apfe1Saft committed:
check tests

Apfe1Saft committed:
check tests

Troitsa commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/FilmDBStorageImpl.java:
```diff
@@ -0,0 +1,133 @@
+        int filmRows = jdbcTemplate.update("insert into films(id,film_name,description,release_date,duration,mpa_id) values(?,?,?,?,?,?)",
+                film.getId(), film.getName(), film.getDescription(), film.getReleaseDate(), film.getDuration(), film.getMpa().getId());
+        for (int id : film.getLikes()) {
+            int likeRows = jdbcTemplate.update("insert into likes(film_id,user_id) values(?,?)", film.getId(), id);
```
эти интовые переменные вроде нигде не используются, точно нужны? или они для дебагга вводились?
